

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Details of operation &mdash; COMET  documentation</title>
  

  
  
    <link rel="shortcut icon" href="_static/COMET_favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Input formatting" href="usage.html" />
    <link rel="prev" title="Background" href="background.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> COMET
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Installation / Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html#direct-installation">Direct Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html#installing-via-a-virtual-environment-recommended">Installing via a Virtual Environment (Recommended)</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html#usage">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html#troubleshooting">Troubleshooting</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="manual.html">User Manual</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="background.html">Background</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Details of operation</a></li>
<li class="toctree-l2"><a class="reference internal" href="usage.html">Input formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="Output.html">Output</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API reference</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">COMET</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="manual.html">User Manual</a> &raquo;</li>
        
      <li>Details of operation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/details.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="details-of-operation">
<h1>Details of operation<a class="headerlink" href="#details-of-operation" title="Permalink to this headline">¶</a></h1>
<p>COMET iterates sequentially through all user-specified clusters, generating statistics/visuals for one at a time.</p>
<div class="contents local topic" id="per-cluster-program-flow">
<p class="topic-title first">Per-cluster program flow</p>
<ul class="simple">
<li><a class="reference internal" href="#read-in-data-from-txt" id="id1">1. Read in data from TXT.</a></li>
<li><a class="reference internal" href="#run-xl-mhg-tests-on-singletons" id="id2">2. Run XL-mHG tests on singletons.</a></li>
<li><a class="reference internal" href="#slide-xl-mhg-cutoff-values" id="id3">3. ‘Slide’ XL-mHG cutoff values.</a></li>
<li><a class="reference internal" href="#generate-discrete-expression-matrix-using-cutoffs" id="id4">4. Generate discrete expression matrix using cutoffs.</a></li>
<li><a class="reference internal" href="#find-cluster-and-population-pair-expression-counts" id="id5">5. Find cluster and population pair expression counts.</a><ul>
<li><a class="reference internal" href="#finding-the-parameters-for-gene-pairs" id="id6">Finding the parameters for gene pairs</a></li>
<li><a class="reference internal" href="#gene-combinations" id="id7">3+ gene combinations</a></li>
</ul>
</li>
<li><a class="reference internal" href="#run-hypergeometric-test-on-pairs-using-counts" id="id8">6. Run hypergeometric test on pairs using counts.</a></li>
<li><a class="reference internal" href="#calculate-true-positive-negative" id="id9">7. Calculate true positive/negative.</a></li>
<li><a class="reference internal" href="#export-statistical-results" id="id10">8. Export statistical results.</a></li>
<li><a class="reference internal" href="#generate-and-export-visualizations" id="id11">9. Generate and export visualizations.</a></li>
</ul>
</div>
<p><img alt="flow" src="_images/flow.jpg" /></p>
<p><em>A simplified version of the program flow.</em></p>
<p><em>First, continuous expression values are converted to discrete via the XL-mHG test. Then, genes are combined and their combinations tests via the hypergeometric test. Finally, these combinations are ranked, assessed, and plotted.</em></p>
<div class="section" id="read-in-data-from-txt">
<h2><a class="toc-backref" href="#id1">1. Read in data from TXT.</a><a class="headerlink" href="#read-in-data-from-txt" title="Permalink to this headline">¶</a></h2>
<p>User-specified data is read in via TXT format (currently, requiring tab delimiters.) See <a class="reference internal" href="usage.html"><span class="doc">Usage</span></a> for details of the format of this data.</p>
<p>There are three components:</p>
<ul class="simple">
<li>Per-cell gene expression values. Each gene must be normalized over cells.</li>
<li>Cluster membership by cell.</li>
<li>Per-cell 2-D visualization coordinates, for plotting.</li>
</ul>
<p>In order to test ‘complement’ genes (i.e. expressed as the absence of a gene rather than the presence of a gene), COMET creates a new complement gene for each gene. These complement genes are treated exactly the same as normal genes, but their expression values are the negative of the expression values of their analogous gene. This allows COMET to find separate mHG cutoff values for genes and their complements.</p>
</div>
<div class="section" id="run-xl-mhg-tests-on-singletons">
<h2><a class="toc-backref" href="#id2">2. Run XL-mHG tests on singletons.</a><a class="headerlink" href="#run-xl-mhg-tests-on-singletons" title="Permalink to this headline">¶</a></h2>
<p>COMET receives as input a gene expression matrix including n cells (sample size) and p genes, as well as a cluster label for each cell. The tool will then proceed to testing, for each cluster C and gene G, whether gene G is a favorable marker candidate for cluster C. A measure of significance will be computed for gene G’s potential as a marker, as well as other informative statistics (such as True Positive and True Negative rates).</p>
<p>COMET will start by pooling all cells that are not in cluster C into a bigger cluster C0. The procedure then relies on the XL-minimal Hypergeometric test (XL-mHG test) to assess whether a gene is a favorable marker for cluster C with respect to cluster C0.  In the following paragraphs, we briefly describe the minima Hypergeometric test (mHG test) and the XL-Hypergeometric test. We refer the reader to Wagner(2017) fo more details on these statistical tests.</p>
<p>The mHG test is a rank-based nonparametric test used to determine gene set enrichment. Cell labels (“1” if the cell belongs to cluster C, and “0” if the cells belongs to cluster C0) are sorted based on absolute gene expression values, resulting in a binary list of cluster labels v. For each given cutoff index i, a Hypergeometric test (HG test) is performed in order to determine enrichment in cluster C cells above the cutoff i. The corresponding HG p-value is computed. The mHG test will then record which cutoff achieved the lowest HG p-value, i.e. the cutoff that achieved the best C cells enrichment in the upper part of the list v. This lowest p-value is the mHG test statistic. A p-value for the mHG test is then computed efficiently using a Dynamic Programming approach.</p>
<p>The XL-mHG test incorporates a slight variation to the mHG test, by including two additional parameters, X and L. L is an upper bound on the cutoff index i (where i=0 represents the top of the list), while X is a percentage value indicating the minimal number of cluster C cells the user wants to capture above the chosen cutoff. Once an optimal cutoff index i* for gene G has been found using the mHG test, the original gene expression vector can be binarized. We consider that G is expressed (value of 1) in all cells above the cutoff i*, and that G is not expressed (value of 0) in all cells below the cutoff. In a sense, X and L can be viewed as parameters controlling the False Positive and False Negative rates. X is precisely a lower bound on the True Positive rate, while L may be tuned to minimize the number of False positives.</p>
<p>Common marker detection practice in single-cell transcriptomics is to compare the cell populations in cluster C and cluster C0 using statistical tests such as the unequal variance t-test (Welch test), the Wilcoxon rank-sum test, the Kolmogorov-Smirnov test, as well as more sophisticated models (zero-inflated Negative Binomial models, hurdle models). COMET will also output the t-test statistic and p-values, but will not use these values for purposes other than informing the user.</p>
</div>
<div class="section" id="slide-xl-mhg-cutoff-values">
<h2><a class="toc-backref" href="#id3">3. ‘Slide’ XL-mHG cutoff values.</a><a class="headerlink" href="#slide-xl-mhg-cutoff-values" title="Permalink to this headline">¶</a></h2>
<p>It may happen that the cutoff index lands at a position in the list where all cells have the same gene expression value. Since we cannot distinguish between cells having the same gene expression value, COMET needs to convert this cutoff index into an absolute  cutoff value t in terms of gene expression (usually expressed in CPM, TPM, or log-transformed TPM). It accomplishes this by ‘sliding up’ the cutoff index in the sorted list of cells until the index encounters a cell with a different gene expression value.</p>
<p>For example, gene expression data often contains many cells that do not express the gene of interest at all. These cells have a gene expression value of 0. In the case that the XL-mHG cutoff index lands on a cell with 0 expression, COMET will ‘slide’ up the cutoff index in the sorted list of cells until it finds a cell with non-zero expression. In other words, COMET will choose a cutoff value between 0 and the lowest non-zero expression value, sliding ‘upwards’ towards higher values. This results in a non-expression (binarized gene expression value of 0) and expression group (binarized gene expression value of 0), where non-expression is exclusively 0-expression.</p>
<p>COMET will always slide ‘upwards’. In the above case of 0-expression, sliding ‘downwards’ and choosing a cutoff below 0 would create an non-expression group of 0 members; no cell expresses below 0! The exception is if COMET reaches the ‘top’ of the list: if it chooses a cutoff value above any cell, the expression group will have zero members. In this case, COMET will turn around and slide ‘downwards’ until it finds a suitable cutoff value. ‘Complement’ genes are treated in the same way, even though their expression values are negative.</p>
</div>
<div class="section" id="generate-discrete-expression-matrix-using-cutoffs">
<h2><a class="toc-backref" href="#id4">4. Generate discrete expression matrix using cutoffs.</a><a class="headerlink" href="#generate-discrete-expression-matrix-using-cutoffs" title="Permalink to this headline">¶</a></h2>
<p>Now that COMET has generated absolute cutoff values for each gene, it converts the gene expression matrix (rows are genes, columns are cells) of continuous gene expression to one of discrete gene expression: where expression for each individual gene G is represented as a ‘1’ for cells that have expression levels above the XL-mHG absolute cutoff value for G, and as a ‘0’ for cells whose G’s expression level is below the XL-mHG absolute cutoff. Binarizing gene expression data is useful to compute statistics such as True Positive and True Negative rates for each gene and each cluster. It also facilitates handling gene combinations.</p>
</div>
<div class="section" id="find-cluster-and-population-pair-expression-counts">
<h2><a class="toc-backref" href="#id5">5. Find cluster and population pair expression counts.</a><a class="headerlink" href="#find-cluster-and-population-pair-expression-counts" title="Permalink to this headline">¶</a></h2>
<p>Once binarized gene expression values have been obtained for each single gene relative to each cluster of interest, one can leverage these results to obtain similar binarized expression values for combinations of genes. For clarity purposes, we specialize this paragraph for combinations of two genes, but this extends beyond this simple case to combinations of any number of genes, or/and their complements. If v1 (resp. v2) represents the binarized gene expression vector for gene A (resp. B), the binarized gene expression vector for the combination (A AND B) will be the vector v1*v2. where * denotes the element-wise product.</p>
<p>Gene combinations are then ranked using the p-values of a Hypergeometric test looking at the number of cells in cluster C that have a binarized (A AND B) expression value of 1. The parameters of the hypergeometric test are as follows:
- Number of cells in cluster C expressing the combination (A AND B);
- Number of cells in cluster C;
- Number of cells expressing the combination (A AND B);
- Total number of cells.
These parameters are usually referred to as <strong>k</strong>, <strong>s</strong>, <strong>M</strong>, and <strong>N</strong>, respectively.</p>
<p>As a note, by default COMET will not consider any single genes with a True-Positive value of less then 15% in any multiple gene marker panels since the True-Positive rate can only decrease with the addition of genes to a marker panel. This also decreases the size of matrix multiplications.</p>
<div class="section" id="finding-the-parameters-for-gene-pairs">
<h3><a class="toc-backref" href="#id6">Finding the parameters for gene pairs</a><a class="headerlink" href="#finding-the-parameters-for-gene-pairs" title="Permalink to this headline">¶</a></h3>
<p>For gene pairs, True Positive and True Negative rates can be computed efficiently using matrix multiplication. To understand this better, let’s take the case of a combination of two genes (A AND B). Let u1 be the vector whose i-th entry is 1 if cell i is a True Positive for gene A (i.e. if cell i belongs to cluster C and expresses gene A), and 0 otherwise. Define u2 similarly for gene B. Then the inner product of u1 with u2 will provide to the number of True Positives for (A AND B), i.e. the number of cells expressing (A AND B) that belong to cluster C. Dividing this inner product by the number of cells in cluster C will yield the True Positive rate. Organizing the True Positive vectors into a gene x cell matrix M, it follows that computing the True Positive rate for gene pairs boils down to matrix multiplication of M with its transpose. The result of this operation is a gene x gene matrix TP, where the (i,j)-th entry TP(i,j) is precisely the True Positive rate corresponding to the combination of gene Gi and gene Gj. This similarly applies to computing True Negative rates.</p>
<p>In this way, COMET finds the <strong>k</strong> and <strong>M</strong> parameters for the hypergeometric test mentioned above, which is used to rank gene combinations in the final output. The <strong>s</strong> and <strong>N</strong> parameters are easily found by counting cluster members.</p>
</div>
<div class="section" id="gene-combinations">
<h3><a class="toc-backref" href="#id7">3+ gene combinations</a><a class="headerlink" href="#gene-combinations" title="Permalink to this headline">¶</a></h3>
<p>Combinations of 3-4 genes are currently implemented and turned on using the -K option. For each given combination of  genes (A AND B AND C), COMET will compute a True Positive rate, True Negative rate as well as a Hypergeometric significance measure, as mentioned above. These computations leverage matrix multiplication. Due to the current computational expense, it can be run in full (if a cluster is available) or on lesser hardware with an abbreviated search space by turning on the heuristics. The main bottleneck is matrix multiplication on computers with smaller memory loads.</p>
<p>For the 3-gene combinations , we construct an ~N^2 x Cells matrix containing 2-gene combinations (A AND B), then multiply by another N x Cells matrix containing gene C. This gives us an expression count matrix for the 3 gene combinations (large) and is then trimmed down to discard gene combinations such as AAA and AAB where there are gene repeats.</p>
</div>
</div>
<div class="section" id="run-hypergeometric-test-on-pairs-using-counts">
<h2><a class="toc-backref" href="#id8">6. Run hypergeometric test on pairs using counts.</a><a class="headerlink" href="#run-hypergeometric-test-on-pairs-using-counts" title="Permalink to this headline">¶</a></h2>
<p>Ranking single genes and gene combinations according to their potential as markers for a given clusters requires defining a unified metric to compare the performance of these different marker candidates. Because each gene combination (where a singleton is considered to be a combination of a single gene) has its own binarized expression vector (where the combinations is said to be expressed for values of 1, and not expressed for values of 0), we can use this expression vector to perform a Hypergeometric test looking at the number of cells in cluster C that have a binarized expression value of 1. This test will produce a p-value, which is then used to rank all different combinations. Intuitively, a lower p-value suggests that the corresponding gene combination is a better candidate marker panel.</p>
</div>
<div class="section" id="calculate-true-positive-negative">
<h2><a class="toc-backref" href="#id9">7. Calculate true positive/negative.</a><a class="headerlink" href="#calculate-true-positive-negative" title="Permalink to this headline">¶</a></h2>
<p>Once COMET has ranked all gene combinations using the corresponding hypergeometric p-value, it reports True Positive and True Negative rates associated with each gene combination, using the matrix products found earlier. As a reminder, True Positive rate is found by dividing the number of expressing cells in the cluster by the total cell count of the cluster; true negative is found by dividing the number of non-expressing cells outside the cluster by the total cell count outside the cluster.</p>
</div>
<div class="section" id="export-statistical-results">
<h2><a class="toc-backref" href="#id10">8. Export statistical results.</a><a class="headerlink" href="#export-statistical-results" title="Permalink to this headline">¶</a></h2>
<p>COMET’s statistical results are exported to multiple CSV files per cluster: some including statistical values for exclusively singletons, and some including combinations. Rows are genes or gene combinations, columns are values associated with that gene/combination.</p>
<p>The columns of the statistical data are:</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">gene</span></code>: the name of the gene.</li>
<li><code class="docutils literal notranslate"><span class="pre">HG_stat</span></code>: the hypergeometric test statistic. Genes are ranked using this value.</li>
<li><code class="docutils literal notranslate"><span class="pre">mHG_pval</span></code>: the statistical significance of the cutoff which the XL-mHG test has chosen.</li>
<li><code class="docutils literal notranslate"><span class="pre">mHG_cutoff_index</span></code>: the index which the XL-mHG has chosen, after sliding.</li>
<li><code class="docutils literal notranslate"><span class="pre">mHG_cutoff_value</span></code>: the gene expression cutoff value corresponding to the chosen index.</li>
<li><code class="docutils literal notranslate"><span class="pre">t_stat</span></code>: the t-test statistic.</li>
<li><code class="docutils literal notranslate"><span class="pre">t_pval</span></code>: the t-test p significance value corresponding to the test statistic.</li>
<li><code class="docutils literal notranslate"><span class="pre">TP</span></code>: the true positive rate.</li>
<li><code class="docutils literal notranslate"><span class="pre">TN</span></code>: the true negative rate.</li>
<li><code class="docutils literal notranslate"><span class="pre">FoldChange</span></code>: the log base 2 fold change of a single gene from the cluster of interest compared to the rest of the cells.</li>
<li><code class="docutils literal notranslate"><span class="pre">init</span> <span class="pre">rank</span></code>: sequential ranking based on <code class="docutils literal notranslate"><span class="pre">HG_stat</span></code>, where lower <code class="docutils literal notranslate"><span class="pre">HG_stat</span></code> is better ranking. No two genes are ranked the same; those with identical hypergeometric statistic values are ranked arbitrarily relative to one another.</li>
<li><code class="docutils literal notranslate"><span class="pre">CCS</span></code>: Cluster Clear Score. This gives us a statistic for the clearing out of other clusters with the addition of the second gene in comparison with the single gene case. The ‘lead gene’ is whichever of the pair has the smallest p-value, then we compute weighted True Negative deltas across the other clusters for the lead gene and the pair. This is only computed for a certain number of the top performing ‘init rank’ gene combinations.</li>
<li><code class="docutils literal notranslate"><span class="pre">rank</span></code>: Final rank. Takes the average of the the init rank and the CCS to determine which gene combinations are the best performing in both tests.</li>
<li><code class="docutils literal notranslate"><span class="pre">Plot</span></code>: Determines whether a gene is plotted (1) or not (0). The purpose is to only graph the top ten of a given gene appearance to avoid very well performing genes from taking over the entire results page.</li>
<li><code class="docutils literal notranslate"><span class="pre">gene</span></code> is replaced by <code class="docutils literal notranslate"><span class="pre">gene_1</span></code> and <code class="docutils literal notranslate"><span class="pre">gene_2</span></code> (combinations of 3+ genes are unimplemented as of yet).</li>
<li><code class="docutils literal notranslate"><span class="pre">mHG_pval</span></code>, <code class="docutils literal notranslate"><span class="pre">mHG_cutoff_index</span></code>, <code class="docutils literal notranslate"><span class="pre">mHG_cutoff_value</span></code>, <code class="docutils literal notranslate"><span class="pre">t_stat</span></code>, and <code class="docutils literal notranslate"><span class="pre">t_pval</span></code> are omitted, since they are irrelevant to non-singletons.</li>
</ul>
</div>
<div class="section" id="generate-and-export-visualizations">
<h2><a class="toc-backref" href="#id11">9. Generate and export visualizations.</a><a class="headerlink" href="#generate-and-export-visualizations" title="Permalink to this headline">¶</a></h2>
<p>For each cluster, COMET generates several PNG file folders containing visualizations of the statistical results. Each folder’s name is prefixed by the cluster name, and includes visuals for only the first few highest ranked genes/combinations. The data files are described in the Output section.</p>
<p>The folders are:</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">(cluster)_pairs_as_singletons</span></code>: compares discrete expression and continuous expression for each gene in a combination on a t-SNE plot.</li>
<li><code class="docutils literal notranslate"><span class="pre">(cluster)_discrete_pairs</span></code>: compares discrete expression of a gene combination with that of its components.</li>
<li><code class="docutils literal notranslate"><span class="pre">(cluster)_pair_TP_TN</span></code>: plots true positive/negative rates of each gene.</li>
<li><code class="docutils literal notranslate"><span class="pre">(cluster)_singleton</span></code>: same as the <code class="docutils literal notranslate"><span class="pre">combined</span></code> plot, but includes only singletons.</li>
<li><code class="docutils literal notranslate"><span class="pre">(cluster)_singleton_TP_TN</span></code>: same as the <code class="docutils literal notranslate"><span class="pre">TP_TN</span></code> plot, but includes only singletons.</li>
<li><code class="docutils literal notranslate"><span class="pre">(cluster)_discrete_trios</span></code>: compares discrete expression of a gene combination with that of its components, but there are three.</li>
</ul>
<div class="toctree-wrapper compound">
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="usage.html" class="btn btn-neutral float-right" title="Input formatting" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="background.html" class="btn btn-neutral float-left" title="Background" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Meromit Singer

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>